api/include/opentelemetry/logs/event_logger.h

```diff
@@ -0,0 +1,94 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#ifdef ENABLE_LOGS_PREVIEW
+
+#  include <chrono>
+#  include <map>
+#  include <vector>
+
+#  include "opentelemetry/common/macros.h"
+#  include "opentelemetry/logs/log_record.h"
+#  include "opentelemetry/logs/logger.h"
+#  include "opentelemetry/logs/logger_type_traits.h"
+#  include "opentelemetry/logs/severity.h"
+#  include "opentelemetry/nostd/shared_ptr.h"
+#  include "opentelemetry/nostd/string_view.h"
+#  include "opentelemetry/nostd/unique_ptr.h"
+#  include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace logs
+{
+/**
+ * Handles event log record creation.
+ **/
+class EventLogger
+{
+public:
+  virtual ~EventLogger() = default;
+
+  /* Returns the name of the logger */
+  virtual const nostd::string_view GetName() noexcept = 0;
+
+  /* Returns the delegate logger of this event logger */
+  virtual nostd::shared_ptr<Logger> GetDelegateLogger() noexcept = 0;
+
+  /**
+   * Emit a event Log Record object
+   *
+   * @param event_name Event name
+   * @param log_record Log record
+   */
+  virtual void EmitEvent(nostd::string_view event_name,
+                         nostd::unique_ptr<LogRecord> &&log_record) noexcept = 0;
+
+  /**
+   * Emit a event Log Record object with arguments
+   *
+   * @param event_name Event name
+   * @tparam args Arguments which can be used to set data of log record by type.
+   *  Severity                                -> severity, severity_text
+   *  string_view                             -> body
+   *  AttributeValue                          -> body
+   *  SpanContext                             -> span_id,tace_id and trace_flags
+   *  SpanId                                  -> span_id
+   *  TraceId                                 -> tace_id
+   *  TraceFlags                              -> trace_flags
+   *  SystemTimestamp                         -> timestamp
+   *  system_clock::time_point                -> timestamp
+   *  KeyValueIterable                        -> attributes
+   *  Key value iterable container            -> attributes
+   *  span<pair<string_view, AttributeValue>> -> attributes(return type of MakeAttributes)
+   */
+  template <class... ArgumentType>
+  void EmitEvent(nostd::string_view event_name, ArgumentType &&... args)
+  {
+    nostd::shared_ptr<Logger> delegate_logger = GetDelegateLogger();
+    if (!delegate_logger)
+    {
+      return;
+    }
+    nostd::unique_ptr<LogRecord> log_record = delegate_logger->CreateLogRecord();
+    if (!log_record)
+    {
+      return;
+    }
+
+    IgnoreTraitResult(
+        detail::LogRecordSetterTrait<typename std::decay<ArgumentType>::type>::template Set(
+            log_record.get(), std::forward<ArgumentType>(args))...);
+
+    EmitEvent(event_name, std::move(log_record));
+  }
```

Comment:
@owent 

So, with the use of variadic templates, type traits, lamda expressions in the implementation, etc ... there is a considerable complexity here.

Because this is in the public API, users trying to figure out how to use this might have some questions and will need a clear documentation in `opentelemetry.io` with examples, and I include myself in "users" here.

This also increases complexity for maintenance, because if this breaks, understanding all the details and providing a fix will take more time, compared to more trivial C++ code.

I assume that the complexity here also saves from some repetitive code that would be also hard to maintain otherwise, so there are plus sides as well.

Overall, the complexity argument does not make the code invalid, so I view it as:

* subjective (people with different experiences with these patterns will have different views)
* and not objective (a given area in the code is technically broken and needs a fix)

To confirm, ok with me to have this code merged to main, knowing fully that I would need to improve my own variadic template skills as a pre requisite if I ever have to make changes here.

LGTM for this part, did not check everything else yet.

Reply:
Thanks for your feedback. Here is my thinking about this complexity.There is a little different between trace and log APIs, with trace, we always need a variable to hold the lifetime of span(Span or Scope) and decide when to call `End()` by RAII, so we can set data field of `Span` one by by one after we create the span.
But in the use case of logs, it's easier for users(I'm also a user of both log and trace.) to use by call just one function, this function is `Logger::Log()` before and `Logger::EmitLogRecord()` now.There are also many data fields we need set to `LogRecord`, log body, trace information, attributes, timestamp and so on.There are three ways to set attributes before(By `const T&`, `std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>>`, `const common::KeyValueIterable &`), and user also can decide whether to set other data fields, log body or trace information. Just consider log body, trace context and attributes, there will be more than `2*2*3+2*2=16` possibilities.Actually, we had more than 10 overloads of `Logger::Log()` before.If we have more options for logs, then the overloads will increase exponentially(I'm trying to add TraceContext now, according to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#logrecord).It will also be hard to maintain these overloads, especially in the future we may add more fields.

I think we can disscuss more deeply if there is any better idea about how to solve this problem.

*The approach of this PR try to use `template <class ValueType> struct LogRecordSetterTrait;` to implement these overloads by type, and also ignore the argument order.We can use `std::decay<ArgumentType>::type` to remove `const`, `reference` and `violate` of `ArgumentType` and use it as the template argument of `template <class ValueType>`, so we can implement any overload action by just write one template specialization for `LogRecordSetterTrait`. And the `template <class ArgumentType> inline static LogRecord *Set(LogRecord *log_record, ArgumentType &&arg) noexcept` use the universal reference so it can accept both left value reference and right value with or without `const` or `violate`.*

Reply:
Thanks @owent for the details.

This seems the best solution then ... and great work.